File: app/src/main/java/com/rahbarbazaar/homadit/android/ui/activities/QRcodeActivity.java
package com.rahbarbazaar.homadit.android.ui.activities;

import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;
//import com.rahbarbazaar.shopper.R;
import com.bumptech.glide.Glide;
import com.rahbarbazaar.homadit.android.R;
import com.rahbarbazaar.homadit.android.controllers.interfaces.BarcodeItemInteraction;
import com.rahbarbazaar.homadit.android.models.barcodlist.Barcode;
import com.rahbarbazaar.homadit.android.models.barcodlist.BarcodeData;
import com.rahbarbazaar.homadit.android.models.barcodlist.BarcodeLoadingState;
import com.rahbarbazaar.homadit.android.models.search_goods.GroupsData;
import com.rahbarbazaar.homadit.android.models.searchable.SearchModel;
import com.rahbarbazaar.homadit.android.models.shopping_memberprize.MemberPrize;
import com.rahbarbazaar.homadit.android.network.Service;
import com.rahbarbazaar.homadit.android.network.ServiceProvider;
import com.rahbarbazaar.homadit.android.ui.fragments.ScanFragment;
import com.rahbarbazaar.homadit.android.utilities.Cache;
import com.rahbarbazaar.homadit.android.utilities.CustomBaseActivity;
import com.rahbarbazaar.homadit.android.utilities.DialogFactory;
import com.rahbarbazaar.homadit.android.utilities.GeneralTools;
import com.rahbarbazaar.homadit.android.utilities.RxBus;
import com.wang.avi.AVLoadingIndicatorView;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class QRcodeActivity extends CustomBaseActivity implements View.OnClickListener, BarcodeItemInteraction {

    LinearLayout linear_return_qrcode, ll_root;
    RelativeLayout rl_home_qrcode;
    Button btn_1, btn_2, btn_3, btn_4, btn_5, btn_6, btn_7, btn_8, btn_finish ,btn_unreadable;

    TextView txt_group_title ;
    ImageView img_icon;

    GeneralTools tools;
    BroadcastReceiver connectivityReceiver = null;
    DialogFactory dialogFactory;

    Disposable disposable = new CompositeDisposable();
    MemberPrize initMemberPrizeLists;

    String description;
    String name;

    AVLoadingIndicatorView avi_qrcode,avi_unreadable ;

    List<SearchModel> searchList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_qrcode1);


        initView();

        //check network broadcast reciever
        tools = GeneralTools.getInstance();
        connectivityReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                tools.doCheckNetwork(QRcodeActivity.this, findViewById(R.id.root_qrcode_scanner));
            }
        };

        disposable = RxBus.MemberPrizeLists.subscribeMemberPrizeLists(result -> {
            if (result instanceof MemberPrize) {
                initMemberPrizeLists = (MemberPrize) result;
            }
        });




//        new Thread (() -> {
//            ScanFragment scanFragment = new ScanFragment();
//            FragmentManager fm = getSupportFragmentManager();
//            FragmentTransaction ft = fm.beginTransaction();
//            ft.replace(R.id.frg_holder, scanFragment);
//            ft.commit();
//        }).start();


        ScanFragment scanFragment = new ScanFragment();
        FragmentManager fm = getSupportFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(R.id.frg_holder, scanFragment);
        ft.commit();


        txt_group_title.setText(Cache.getString(this,"selectedGroupTitle"));
        String url = Cache.getString(this,"selectedGroupIcon");
        Glide
                .with(this)
                .load(url)
                .centerCrop()
                .placeholder(R.color.blue)
                .into(img_icon);

        avi_qrcode.hide();
    }



    private void initView() {

        linear_return_qrcode = findViewById(R.id.linear_return_qrcode);
        rl_home_qrcode = findViewById(R.id.rl_home_qrcode);
        txt_group_title  = findViewById(R.id.txt_group_title);
        img_icon  = findViewById(R.id.img_icon);
        ll_root = findViewById(R.id.root_qrcode_scanner);
        btn_1 = findViewById(R.id.btn1_register_barcode);
        btn_2 = findViewById(R.id.btn2_register_barcode);
        btn_3 = findViewById(R.id.btn3_register_barcode);
        btn_4 = findViewById(R.id.btn4_register_barcode);
        btn_5 = findViewById(R.id.btn5_register_barcode);
        btn_6 = findViewById(R.id.btn6_register_barcode);
        btn_7 = findViewById(R.id.btn7_register_barcode);
        btn_8 = findViewById(R.id.btn8_register_barcode);
        btn_finish = findViewById(R.id.btn_finish_purchase_qrcode);
        avi_qrcode = findViewById(R.id.avi_qrcode);
        avi_unreadable = findViewById(R.id.avi_unreadable);

        btn_unreadable = findViewById(R.id.btn_unreadable);




//        btn_1.setText(initMemberPrizeLists.data.categories.get(0).name);
//        btn_2.setText(initMemberPrizeLists.data.categories.get(1).name);
//        btn_3.setText(initMemberPrizeLists.data.categories.get(2).name);
//        btn_4.setText(initMemberPrizeLists.data.categories.get(3).name);
//        btn_5.setText(initMemberPrizeLists.data.categories.get(4).name);
//        btn_6.setText(initMemberPrizeLists.data.categories.get(5).name);
//        btn_7.setText(initMemberPrizeLists.data.categories.get(6).name);
//        btn_8.setText(initMemberPrizeLists.data.categories.get(7).name);


        btn_finish.setOnClickListener(this);
        linear_return_qrcode.setOnClickListener(this);
        rl_home_qrcode.setOnClickListener(this);
//        btn_1.setOnClickListener(this);
//        btn_2.setOnClickListener(this);
//        btn_3.setOnClickListener(this);
//        btn_4.setOnClickListener(this);
//        btn_5.setOnClickListener(this);
//        btn_6.setOnClickListener(this);
//        btn_7.setOnClickListener(this);
//        btn_8.setOnClickListener(this);
        btn_unreadable.setOnClickListener(this);

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {

            case R.id.linear_return_qrcode:
                Intent intent = new Intent(QRcodeActivity.this,GroupGoodsActivity.class);
                intent.putExtra("new_register","repeat");
                startActivity(intent);

                finish();
                break;

            case R.id.rl_home_qrcode:
            case R.id.btn_finish_purchase_qrcode:
                startActivity(new Intent(QRcodeActivity.this, MainActivity.class));
                finish();
                break;

//            case R.id.btn1_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(0).description;
//                name = initMemberPrizeLists.data.categories.get(0).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn2_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(1).description;
//                name = initMemberPrizeLists.data.categories.get(1).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn3_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(2).description;
//                name = initMemberPrizeLists.data.categories.get(2).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn4_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(3).description;
//                name = initMemberPrizeLists.data.categories.get(3).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn5_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(4).description;
//                name = initMemberPrizeLists.data.categories.get(4).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn6_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(5).description;
//                name = initMemberPrizeLists.data.categories.get(5).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn7_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(6).description;
//                name = initMemberPrizeLists.data.categories.get(6).name;
//                showInfoDialog(description, name);
//                break;
//
//            case R.id.btn8_register_barcode:
//                description = initMemberPrizeLists.data.categories.get(7).description;
//                name = initMemberPrizeLists.data.categories.get(7).name;
//                showInfoDialog(description, name);
//                break;


            case R.id.btn_unreadable:
//                getCategoryList();
                Intent intent2 = new Intent(QRcodeActivity.this,PurchasedItemActivity.class);
                intent2.putExtra("unreadable_barcode","unreadable_barcode");
                startActivity(intent2);
                break;


        }
    }

    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
    public void onEventMainThread(BarcodeLoadingState state) {

        if (state.getState().equals("show_loading")) {
            avi_qrcode.show();
        } else if (state.getState().equals("stop_loading")) {
            avi_qrcode.hide();
        }
    }

    @Subscribe
    public void onEvent(Barcode barcode) {
//        Toast.makeText(this, ""+barcode.getData().get(0).getDecription(), Toast.LENGTH_SHORT).show();
        showBarcodeListDialog(barcode);
    }

    private void showInfoDialog(String description, String name) {

        dialogFactory = new DialogFactory(QRcodeActivity.this);
        dialogFactory.createQrcodeInfoBtnsDialog(new DialogFactory.DialogFactoryInteraction() {
            @Override
            public void onAcceptButtonClicked(String... params) {

            }

            @Override
            public void onDeniedButtonClicked(boolean bool) {
            }
        }, ll_root, description, name);

    }

    private void showBarcodeListDialog(Barcode barcode) {
        DialogFactory dialogFactory = new DialogFactory(QRcodeActivity.this);
        dialogFactory.createBarcodeResultListDialog(new DialogFactory.DialogFactoryInteraction() {
            @Override
            public void onAcceptButtonClicked(String... params) {
                Intent intent = new Intent(QRcodeActivity.this, PurchasedItemActivity.class);
                intent.putExtra("unreadable_barcode", "unreadable_barcode");
                startActivity(intent);
                finish();
            }

            @Override
            public void onDeniedButtonClicked(boolean bool) {
            }
        }, ll_root, barcode, this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        EventBus.getDefault().register(this);
        registerReceiver(connectivityReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
    }


    @Override
    public void barcodeListOnClicked(@NotNull BarcodeData model, int position, @NotNull Barcode barcode, @NotNull AlertDialog dialog) {
        Intent intent = new Intent(QRcodeActivity.this, PurchasedItemActivity.class);
        intent.putExtra("position", position);
        intent.putExtra("barcodeList", barcode);
        intent.putExtra("suggested_price",model.getPrice());
        intent.putExtra("product_id",barcode.getData().get(position).getId());
        intent.putExtra("mygroup",barcode.getData().get(position).getMygroup());
//        intent.putExtra("min_price",barcode.getData().get(position).getMinPrice());
//        intent.putExtra("max_price",barcode.getData().get(position).getMaxPrice());
//        intent.putExtra("max_amount",barcode.getData().get(position).getMaxAmount());

        startActivity(intent);
        overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
        finish();

    }

    @Override
    public void onStop() {
        super.onStop();
        EventBus.getDefault().unregister(this);
        unregisterReceiver(connectivityReceiver);
        disposable.dispose();
    }


    private void getCategoryList() {

        avi_unreadable.setVisibility(View.VISIBLE);
        btn_unreadable.setVisibility(View.GONE);

        Service service = new ServiceProvider(this).getmService();
        Call<GroupsData> call = service.getCategorySpnData();
        call.enqueue(new Callback<GroupsData>() {
            @Override
            public void onResponse(Call<GroupsData> call, Response<GroupsData> response) {
                if (response.code() == 200) {

                    avi_unreadable.setVisibility(View.GONE);
                    btn_unreadable.setVisibility(View.VISIBLE);

                    searchList = new ArrayList<>();

                    for (int i = 0; i < response.body().getData().size(); i++) {
                        searchList.add(new SearchModel(response.body().getData().get(i).getTitle(),
                                response.body().getData().get(i).getId()));
                    }

                    RxBus.GroupGoodsList.publishGroupGoodsList(response.body());
                    startActivity(new Intent(QRcodeActivity.this,GroupGoodsActivity.class));


                } else {
                    Toast.makeText(QRcodeActivity.this, "" + getResources().getString(R.string.serverFaield), Toast.LENGTH_SHORT).show();
                    avi_unreadable.setVisibility(View.GONE);
                    btn_unreadable.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onFailure(Call<GroupsData> call, Throwable t) {
                Toast.makeText(QRcodeActivity.this, "" + getResources().getString(R.string.connectionFaield), Toast.LENGTH_SHORT).show();
                avi_unreadable.setVisibility(View.GONE);
                btn_unreadable.setVisibility(View.VISIBLE);
            }
        });

    }


    @Override
    public void onBackPressed() {
//        super.onBackPressed();

        Intent intent = new Intent(QRcodeActivity.this,GroupGoodsActivity.class);
        intent.putExtra("new_register","repeat");
        startActivity(intent);

        finish();
    }
}

File: app/src/main/java/com/rahbarbazaar/shopper/ui/activities/QRcodeActivity.kt
package com.rahbarbazaar.shopper.ui.activities

import android.Manifest
import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import com.rahbarbazaar.shopper.R
import com.rahbarbazaar.shopper.models.api_error.ErrorUtils
import com.rahbarbazaar.shopper.models.barcodlist.Barcode
import com.rahbarbazaar.shopper.models.search_goods.GroupsData
import com.rahbarbazaar.shopper.models.shopping_memberprize.MemberPrize
import com.rahbarbazaar.shopper.network.ServiceProvider
import com.rahbarbazaar.shopper.utilities.*
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.activity_qrcode.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class QRcodeActivity : CustomBaseActivity(), View.OnClickListener {

    private var connectivityReceiver: BroadcastReceiver? = null
    lateinit var initMemberPrizeLists: MemberPrize
    var disposable: Disposable = CompositeDisposable()

    companion object {
        lateinit var ResultScan: String
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_qrcode)

        //check network broadcast reciever
        val tools = GeneralTools.getInstance()
        connectivityReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                tools.doCheckNetwork(this@QRcodeActivity, findViewById<View>(R.id.rl_root))
            }
        }

        initMemberPrizeLists = MemberPrize()
        disposable = RxBus.MemberPrizeLists.subscribeMemberPrizeLists() { result ->
            if (result is MemberPrize) {
                initMemberPrizeLists = result
            }
        }

        btn_choose_scanner.setOnClickListener(this)
        btn_register_barcode.setOnClickListener(this)
        rl_home_qrcode.setOnClickListener(this)
        linear_return_qrcode.setOnClickListener(this)
//        btn_barcode_search.setOnClickListener(this)
        rl_barcode_search.setOnClickListener(this)
        btn_direct_register.setOnClickListener(this)
        btn_finish_purchased.setOnClickListener(this)
        rl_info_registerbarcode.setOnClickListener(this)


        // state can be null
        val state: String? = intent.getStringExtra("static_barcode")
        if (state == "static_barcode") {
            edt_barcode.setText("")
        } else {
            edt_barcode?.setText(ResultScan)
        }

        txt_list_registable.text = initMemberPrizeLists.data.categories


        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)




        edt_barcode

        // event on done keyboard
        edt_barcode.setOnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                getListOfProducts()
                closeKeyboard()
                return@setOnEditorActionListener true
            }
            false
        }


//        // Spinner click listener
//        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
//            public void onItemSelected(AdapterView<?> parent, View view,
//                    int position, long id) {
//
//                // On selecting a spinner item
//                String item = parent.getItemAtPosition(position).toString();
//
//
//                // Showing selected spinner item
//                Toast.makeText(parent.getContext(), "Selected: " + item, Toast.LENGTH_LONG).show();
//
//            }
//
//            @Override
//            public void onNothingSelected(AdapterView<?> parent) {
//
//            }
//        });


    }

    private fun closeKeyboard() {
        val inputMethodManager = getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        inputMethodManager.hideSoftInputFromWindow(
                (currentFocus).windowToken, 0)
    }


    override fun onClick(view: View) {
        when (view.id) {

            R.id.btn_choose_scanner -> {

                if (checkCameraPermission()) {
                    startActivity(Intent(this@QRcodeActivity, ScanActivity::class.java))
                    overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
                    finish()
                } else {
                    requestCameraPermission()
                }
            }


            R.id.btn_register_barcode -> {
                getListOfProducts()
            }

            R.id.rl_home_qrcode -> {
                startActivity(Intent(this@QRcodeActivity, MainActivity::class.java))
                overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
                finish()
            }

            R.id.linear_return_qrcode -> {
                finish()
            }

            R.id.rl_barcode_search -> {

                getGroupSpnList()
            }

            R.id.btn_direct_register -> {
                val intent = Intent(this@QRcodeActivity, PurchasedItemsActivity::class.java)
                intent.putExtra("no_searchedList", "no_searchedList")
                startActivity(intent)
                overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
            }

            R.id.btn_finish_purchased -> {
                startActivity(Intent(this@QRcodeActivity,MainActivity::class.java))
                overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
                Toast.makeText(this@QRcodeActivity,""+resources.getString(R.string.finish_purchase_info2),Toast.LENGTH_LONG).show()
                finish()
            }

            R.id. rl_info_registerbarcode -> {

                showRegisterBarcodeInfoDialog()
            }


        }
    }

    private fun showRegisterBarcodeInfoDialog() {

        val dialogFactory = DialogFactory(this)
        dialogFactory.createRegisterBarcodeInfoDialog(object : DialogFactory.DialogFactoryInteraction {
            override fun onAcceptButtonClicked(vararg strings: String?) {
            }

            override fun onDeniedButtonClicked(cancel_dialog: Boolean) {

            }

        }, rl_root)
    }

    private fun getGroupSpnList() {

        btn_barcode_search.visibility = View.GONE
        avi_barcode_search.visibility = View.VISIBLE

        val service = ServiceProvider(this).getmService()
        val call = service.groupsSpnData
        call.enqueue(object : Callback<GroupsData> {
            override fun onResponse(call: Call<GroupsData>, response: Response<GroupsData>) {
                if (response.code() == 200) {

                    var groupsData = GroupsData()
                    groupsData = response.body()!!

                    val intent = Intent(this@QRcodeActivity, SearchActivity::class.java)
                    intent.putExtra("groupsData", groupsData)
                    startActivity(intent)
//                    RxBus.GroupsSpnData.publishGroupsSpnData(groupsData)
//                    startActivity(Intent(this@QRcodeActivity, SearchActivity::class.java))
                    overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
                    btn_barcode_search.visibility = View.VISIBLE
                    avi_barcode_search.visibility = View.GONE
                    finish()

                } else {
                    Toast.makeText(this@QRcodeActivity, resources.getString(R.string.serverFaield), Toast.LENGTH_SHORT).show()
                    btn_barcode_search.visibility = View.VISIBLE
                    avi_barcode_search.visibility = View.GONE

                }
            }

            override fun onFailure(call: Call<GroupsData>, t: Throwable) {
                Toast.makeText(this@QRcodeActivity, resources.getString(R.string.connectionFaield), Toast.LENGTH_SHORT).show()
                btn_barcode_search.visibility = View.VISIBLE
                avi_barcode_search.visibility = View.GONE

            }
        })

    }

    private fun checkCameraPermission(): Boolean {
        return ContextCompat.checkSelfPermission(this@QRcodeActivity, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestCameraPermission() {
        ActivityCompat.requestPermissions(this@QRcodeActivity, arrayOf(Manifest.permission.CAMERA), 33)
    }

    private fun getListOfProducts() {

        showAvi()

        val barcode = edt_barcode.text.toString()
        val service = ServiceProvider(this).getmService()
        val call = service.getBarcodeList(barcode)
        call.enqueue(object : Callback<Barcode> {
            override fun onResponse(call: Call<Barcode>, response: Response<Barcode>) {
                if (response.code() == 200) {

                    var barcode = Barcode()
                    barcode = response.body()!!

                    RxBus.BarcodeList.publishBarcodeList(barcode)
//                  startActivity(Intent(this@QRcodeActivity, BarcodeListActivity::class.java))
                    val intent = Intent(this@QRcodeActivity, BarcodeListActivity::class.java)
                    intent.putExtra("barcode", edt_barcode.text.toString())
                    startActivity(intent)
                    overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
                    showbtn()
                    finish()

                } else if (response.code() == 422) {
                    showbtn()
                    val apiError = ErrorUtils.parseError422(response)
                    if (apiError.errors.barcode != null) {

                        var builderBarcode = StringBuilder()
                        for (a in apiError.errors.barcode) {
                            builderBarcode.append("$a ")
                        }
                        Toast.makeText(this@QRcodeActivity, "" + builderBarcode, Toast.LENGTH_SHORT).show()

                    }


                } else if (response.code() == 204) {
                    val intent = Intent(this@QRcodeActivity, PurchasedItemsActivity::class.java)
                    intent.putExtra("no_product", "no_product")
                    intent.putExtra("barcode", edt_barcode.text.toString())
                    startActivity(intent)
                    overridePendingTransition(R.anim.slide_in, R.anim.slide_out)

                    showbtn()
                } else {
                    Toast.makeText(this@QRcodeActivity, resources.getString(R.string.serverFaield), Toast.LENGTH_SHORT).show()
                    showbtn()
                }
            }

            override fun onFailure(call: Call<Barcode>, t: Throwable) {
                Toast.makeText(this@QRcodeActivity, resources.getString(R.string.connectionFaield), Toast.LENGTH_SHORT).show()
                showbtn()
            }
        })
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == 33) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startActivity(Intent(this@QRcodeActivity, ScanActivity::class.java))
            } else {
                Toast.makeText(this, "نیاز به اجازه ی دسترسی دوربین", Toast.LENGTH_SHORT).show()
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun showbtn() {
        btn_register_barcode.visibility = View.VISIBLE
        avi_register_barcode.visibility = View.GONE
    }

    private fun showAvi() {
        btn_register_barcode.visibility = View.GONE
        avi_register_barcode.visibility = View.VISIBLE
    }

    override fun onResume() {
        super.onResume()
        registerReceiver(connectivityReceiver, IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION))

        // to delete BarcodeEditText while register successfully
        val barcode_state: String? = Cache.getString(this@QRcodeActivity, "barcode_registered")
        if (barcode_state == "barcode_registered") {
            edt_barcode.setText("")
            Cache.setString(this@QRcodeActivity, "barcode_registered", "")
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(connectivityReceiver)
        disposable.dispose()
    }


}